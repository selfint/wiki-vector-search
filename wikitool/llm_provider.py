from typing import Protocol, TypeVar

T = TypeVar("T")


class LLMProvider(Protocol[T]):
    def embed_queries(self, texts: list[str] | str) -> T:
        ...

    def embed_corpus(self, texts: list[str] | str) -> T:
        ...

    def chunk(self, text: str, size: int, overlap: int) -> list[str]:
        ...

    def search(
        self,
        query_embeddings: T,
        corpus_embeddings: T,
        top_k: int,
    ) -> list[list[int]]:
        ...
