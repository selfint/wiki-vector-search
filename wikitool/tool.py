from typing import Callable, TypeAlias
from transformers import PreTrainedTokenizerFast
import collections
from functools import partial


Tree: TypeAlias = dict[str, tuple[str, "Tree"]]


def _tree() -> collections.defaultdict:
    return collections.defaultdict(_tree)


def build_tree(sections, tree: Tree | None = None) -> Tree:
    if tree is None:
        tree = _tree()

    for s in sections:
        tree[s.title] = (s.text, build_tree(s.sections))

    return dict(tree)


def chunk_tokens(
    tokens: list[str],
    chunk_size: int,
    chunk_overlap: int,
    is_subword: Callable[[str], bool],
) -> list[list[str]]:
    chunks = []

    while len(tokens) > 0:
        # reduce chunk size to not end with subword
        chunk_len = chunk_size
        while (chunk_len + 1 < len(tokens)) and is_subword(tokens[chunk_len + 1]):
            chunk_len -= 1
        assert chunk_len > 0, "got empty chunk"

        chunk_tokens = tokens[:chunk_len]
        chunks.append(chunk_tokens)

        new_start = chunk_len
        if new_start > len(tokens):
            break

        # try to reduce overlap to not cut subwords
        overlap = chunk_overlap if chunk_overlap < chunk_len else 0
        while is_subword(tokens[new_start - overlap]):
            overlap -= 1
            if overlap == 0:
                print(
                    "failed to prevent subword cut:",
                    tokens[new_start - 1],
                    tokens[new_start],
                )
                break

        tokens = tokens[new_start - overlap :]

    return chunks


def build_text_chunks(
    text: str,
    title: str,
    title_path: list[str],
    tokenizer: PreTrainedTokenizerFast,
    chunk_size: int,
    chunk_overlap: int = 0,
) -> list[dict]:
    # build header
    header = (
        "\n".join(f"{'#' * (i + 1)} {t}" for i, t in enumerate(title_path + [title]))
        + "\n"
    )

    # split tokens into chunk sized chunks
    text_chunk_size = chunk_size - len(tokenizer.tokenize(header))
    text_chunks_tokens = chunk_tokens(
        tokenizer.tokenize(text),
        text_chunk_size,
        chunk_overlap,
        is_subword=lambda t: not t.startswith("▁"),
    )

    # build chunks
    chunks = [
        {
            "text": header + tokenizer.decoder.decode(text_chunk),
            "meta": {"title": title, "title_path": title_path},
        }
        for text_chunk in text_chunks_tokens
    ]

    return chunks


def build_tree_chunks(
    tree: Tree,
    title_path: list[str],
    tokenizer: PreTrainedTokenizerFast,
    chunk_size: int,
    chunk_overlap: int = 0,
) -> list[dict]:
    assert chunk_size > chunk_overlap * 2, f"overlap must be < {chunk_size // 2=}"
    _build_text_chunks = partial(
        build_text_chunks,
        tokenizer=tokenizer,
        chunk_size=chunk_size,
        chunk_overlap=chunk_overlap,
    )

    chunks = []
    title_path = title_path or []

    for k, (text, subtrees) in tree.items():
        chunks += _build_text_chunks(text, k, title_path)

        # build subtree chunks
        for subtitle, (subtext, subtree) in subtrees.items():
            chunks += _build_text_chunks(subtext, subtitle, title_path + [k])
            chunks += build_tree_chunks(
                subtree,
                title_path,
                tokenizer,
                chunk_size,
                chunk_overlap,
            )

    return chunks
