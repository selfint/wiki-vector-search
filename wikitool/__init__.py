from typing import Annotated
from typing_extensions import Doc
from wikipediaapi import Wikipedia
from .llm_provider import LLMProvider

__version__ = "0.1.0"


class WikiTool:
    def __init__(
        self,
        client: Annotated[Wikipedia, Doc("Wikipedia client")],
        llm: Annotated[LLMProvider, Doc("LLM Provider")],
    ) -> None:
        """
        Wiki Tool.

        Example:
            Create a WikiTool object
            ```pycon
            >>> from wikipediaapi import Wikipedia
            >>> from wikitool import WikiTool
            >>> from wikitool.extras.st_provider import STProvider
            >>> tool = WikiTool(
            ...         client=Wikipedia("user_agent"),
            ...         llm=STProvider("thenlper/gte-small"),
            ...        )

            ```
        """
        self._client = client
        self._llm = llm_provider
