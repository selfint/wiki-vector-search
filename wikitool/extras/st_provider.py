from typing import TypeAlias

import torch
from numpy.typing import NDArray
from sentence_transformers import SentenceTransformer
from sentence_transformers.util import semantic_search
from torch import Tensor
from transformers import AutoTokenizer, PreTrainedTokenizerFast

from ..llm_provider import LLMProvider

T: TypeAlias = Tensor | NDArray


class STProvider(LLMProvider[T]):
    def __init__(
        self,
        model: str,
        instruction: str | None = None,
        device: str | None = None,
        half: bool = False,
    ) -> None:
        self._device = device or (
            "cuda"
            if torch.cuda.is_available()
            else "mps"
            if torch.backends.mps.is_available()
            else "cpu"
        )

        self._model = SentenceTransformer(model, device=self._device)
        if half:
            self._model = self._model.half()
        self._model.compile()

        self._tokenizer = AutoTokenizer.from_pretrained(model)

        self._instruction = instruction

        self._normalize = True

    def embed_corpus(self, texts: list[str] | str) -> T:
        return self._model.encode(
            texts,
            device=self._device,
            normalize_embeddings=self._normalize,
            convert_to_numpy=self._device == "cpu",
            convert_to_tensor=self._device != "cpu",
        )  # type: ignore

    def embed_queries(self, texts: list[str] | str) -> T:
        if self._instruction is not None:
            texts = [self._instruction + t for t in texts]

        return self._model.encode(
            texts,
            device=self._device,
            normalize_embeddings=self._normalize,
            convert_to_numpy=self._device == "cpu",
            convert_to_tensor=self._device != "cpu",
        )  # type: ignore

    def chunk(self, text: str, size: int, overlap: int) -> list[str]:
        if len(text) <= 1:
            return [text]

        tokenizer: PreTrainedTokenizerFast = self._pipe.tokenizer  # type: ignore

        tokens = tokenizer.tokenize(
            text,
            max_length=size,
            stride=overlap,
            truncation=True,
            add_special_tokens=False,
            return_overflowing_tokens=True,
            return_offsets_mapping=True,
        )

        token_offsets: list[list[tuple[int, int]]] = tokens["offset_mapping"]  # type: ignore
        chunk_offsets = [(offsets[0][0], offsets[-1][1]) for offsets in token_offsets]

        chunks = [text[start:end] for start, end in chunk_offsets]

        return chunks

    def search(
        self,
        query_embeddings: T,
        corpus_embeddings: T,
        top_k: int,
    ) -> list[list[int]]:
        hits = semantic_search(
            query_embeddings=query_embeddings,  # type: ignore
            corpus_embeddings=corpus_embeddings,  # type: ignore
            top_k=top_k,
        )

        return [[h["corpus_id"] for h in query_hits] for query_hits in hits]
