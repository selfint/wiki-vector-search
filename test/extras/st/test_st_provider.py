from functools import lru_cache

import pytest

from wikitool.extras.st_provider import STProvider
from wikitool.llm_provider import LLMProvider


@lru_cache(maxsize=1)
def _st_provider() -> STProvider:
    """Only load model once"""

    return STProvider("thenlper/gte-small", device="cpu")


@pytest.fixture
def st_provider() -> LLMProvider:
    return _st_provider()


def test_embed(st_provider, snapshot):
    embedding = st_provider.embed_corpus(["this is a test", "and another test"])

    # snapshot first 2 digits to bypass float issues
    snap = [[int(a * 100) for a in b] for b in embedding]

    assert snap == snapshot


def test_chunk(st_provider, snapshot):
    text = "The quick brown fox jumps over the lazy dog"

    chunks = st_provider.chunk(text, 3, 1)

    assert chunks == snapshot


def test_search(st_provider, snapshot):
    corpus = [
        "The name of the dog is Kevin",
        "The house is blue",
        "It is Tuesday",
    ]

    queries = [
        "What is the pet's name?",
        "What color is the house?",
        "What day is it?",
    ]

    top_k = 1
    results = st_provider.search(
        st_provider.embed_queries(queries),
        st_provider.embed_corpus(corpus),
        top_k=top_k,
    )

    assert [len(r) for r in results] == [top_k] * len(results)
    results = [[corpus[i] for i in r] for r in results]

    assert list(zip(queries, results)) == snapshot
