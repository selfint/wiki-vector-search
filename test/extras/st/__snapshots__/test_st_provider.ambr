# serializer version: 1
# name: test_chunk
  list([
    '''
      # Title
      ## Title2
      The quick brown fox
    ''',
    '''
      # Title
      ## Title2
      fox jumps over the
    ''',
    '''
      # Title
      ## Title2
      the lazy dog
    ''',
  ])
# ---
# name: test_embed
  list([
    list([
      -8,
      1,
      3,
      2,
      1,
      6,
      9,
      0,
      0,
      -2,
      0,
      -5,
      1,
      3,
      -2,
      -3,
      1,
      -2,
      -2,
      4,
      7,
      2,
      -5,
      -3,
      4,
      2,
      0,
      -5,
      -4,
      -19,
      0,
      -7,
      4,
      -5,
      2,
      -3,
      -4,
      1,
      -3,
      2,
      5,
      -1,
      -1,
      -4,
      -1,
      -2,
      -2,
      -1,
      1,
      -3,
      1,
      -4,
      1,
      -1,
      1,
      5,
      5,
      1,
      3,
      2,
      2,
      3,
      -20,
      9,
      4,
      2,
      -4,
      -2,
      4,
      1,
      -4,
      3,
      3,
      7,
      1,
      -3,
      -1,
      -7,
      2,
      1,
      -2,
      -5,
      0,
      0,
      -1,
      -6,
      0,
      0,
      6,
      2,
      -1,
      0,
      -2,
      1,
      -3,
      -4,
      3,
      0,
      -1,
      25,
      -7,
      2,
      1,
      -4,
      0,
      -4,
      1,
      -6,
      -2,
      -2,
      0,
      -3,
      7,
      -3,
      0,
      0,
      5,
      4,
      -1,
      2,
      0,
      1,
      1,
      -7,
      7,
      -7,
      6,
      11,
      2,
      1,
      4,
      -3,
      -7,
      -1,
      -2,
      2,
      0,
      1,
      1,
      0,
      -3,
      -12,
      0,
      -8,
      -7,
      4,
      -3,
      7,
      -4,
      0,
      0,
      4,
      -4,
      -3,
      1,
      0,
      2,
      3,
      -5,
      3,
      -1,
      -4,
      -6,
      6,
      0,
      -5,
      -3,
      0,
      2,
      -3,
      5,
      2,
      -7,
      1,
      1,
      0,
      -1,
      2,
      0,
      0,
      6,
      -2,
      -3,
      6,
      4,
      0,
      -1,
      -5,
      3,
      3,
      -1,
      6,
      -7,
      0,
      -3,
      -3,
      0,
      -1,
      2,
      -1,
      5,
      0,
      1,
      5,
      0,
      8,
      0,
      -1,
      4,
      4,
      -3,
      0,
      0,
      0,
      -2,
      -1,
      1,
      1,
      0,
      6,
      0,
      -7,
      0,
      -24,
      0,
      2,
      0,
      7,
      -3,
      1,
      0,
      4,
      2,
      6,
      0,
      -1,
      2,
      0,
      4,
      1,
      -1,
      1,
      2,
      0,
      0,
      1,
      -3,
      7,
      -4,
      22,
      9,
      1,
      -1,
      3,
      1,
      0,
      -9,
      4,
      5,
      -1,
      0,
      -6,
      -3,
      -4,
      0,
      0,
      -8,
      -1,
      -1,
      0,
      4,
      -7,
      0,
      3,
      -8,
      0,
      4,
      2,
      -2,
      -11,
      -4,
      -2,
      0,
      0,
      2,
      4,
      -1,
      5,
      2,
      -2,
      -6,
      5,
      -1,
      2,
      4,
      -1,
      -2,
      3,
      2,
      3,
      -3,
      0,
      -2,
      1,
      0,
      0,
      2,
      2,
      1,
      3,
      -2,
      5,
      -5,
      0,
      3,
      -5,
      -8,
      0,
      0,
      -27,
      4,
      0,
      2,
      -5,
      0,
      2,
      3,
      -10,
      1,
      4,
      5,
      3,
      -1,
      0,
      5,
      7,
      -5,
      3,
      -7,
      0,
      0,
      19,
      -3,
      3,
      2,
      2,
      0,
      7,
      -2,
      6,
      -1,
      5,
      -6,
      2,
      2,
      -3,
      6,
      2,
      -1,
      0,
      -1,
      -4,
      0,
      14,
      -1,
      -2,
      -5,
      -3,
      0,
      -3,
      -2,
      0,
      0,
      5,
      4,
      -3,
      -5,
      -2,
      -4,
      0,
      -4,
      1,
      9,
      3,
    ]),
    list([
      -8,
      2,
      4,
      0,
      2,
      4,
      6,
      2,
      0,
      -5,
      0,
      -9,
      1,
      2,
      -1,
      2,
      -2,
      0,
      -6,
      4,
      4,
      1,
      -3,
      -4,
      3,
      4,
      -3,
      -5,
      -3,
      -21,
      0,
      -7,
      0,
      -6,
      1,
      -5,
      -4,
      1,
      -3,
      2,
      2,
      -2,
      -3,
      -4,
      -2,
      -2,
      -2,
      -2,
      5,
      -2,
      3,
      -5,
      3,
      0,
      2,
      1,
      4,
      2,
      0,
      4,
      2,
      3,
      -21,
      5,
      7,
      2,
      -7,
      -2,
      3,
      8,
      0,
      2,
      2,
      7,
      0,
      -1,
      -2,
      -4,
      0,
      0,
      -3,
      -2,
      1,
      0,
      0,
      -2,
      -2,
      0,
      4,
      2,
      0,
      -2,
      1,
      2,
      -5,
      -3,
      1,
      0,
      -2,
      26,
      -7,
      0,
      1,
      -4,
      3,
      -2,
      -2,
      -5,
      -4,
      -1,
      3,
      -3,
      5,
      -3,
      0,
      4,
      5,
      0,
      -1,
      1,
      0,
      0,
      4,
      -9,
      4,
      -7,
      5,
      11,
      0,
      0,
      4,
      -6,
      -4,
      0,
      -1,
      1,
      0,
      0,
      5,
      0,
      -3,
      -11,
      -1,
      -9,
      -4,
      7,
      -1,
      6,
      -3,
      2,
      -1,
      6,
      -1,
      -1,
      3,
      0,
      4,
      3,
      -1,
      2,
      -1,
      -4,
      -7,
      6,
      1,
      -4,
      -3,
      2,
      1,
      -2,
      7,
      2,
      -8,
      -2,
      1,
      -2,
      -3,
      4,
      0,
      3,
      3,
      0,
      -5,
      5,
      4,
      -2,
      0,
      -7,
      2,
      5,
      -1,
      5,
      -5,
      3,
      -2,
      -4,
      -2,
      -1,
      -1,
      -4,
      3,
      -1,
      1,
      4,
      3,
      6,
      2,
      -4,
      5,
      1,
      -3,
      0,
      5,
      -2,
      -1,
      0,
      0,
      1,
      0,
      6,
      0,
      -4,
      3,
      -24,
      -2,
      2,
      -1,
      5,
      -4,
      2,
      2,
      2,
      5,
      3,
      0,
      -3,
      3,
      1,
      4,
      2,
      1,
      1,
      0,
      0,
      4,
      1,
      -3,
      5,
      -3,
      19,
      7,
      2,
      -6,
      4,
      4,
      2,
      -9,
      7,
      4,
      0,
      2,
      -6,
      -5,
      -4,
      1,
      1,
      -5,
      0,
      -2,
      0,
      3,
      -4,
      1,
      2,
      -5,
      0,
      -2,
      2,
      -4,
      -11,
      -2,
      -2,
      2,
      0,
      0,
      2,
      -2,
      6,
      2,
      -2,
      -4,
      2,
      -1,
      1,
      3,
      0,
      -3,
      5,
      1,
      4,
      -5,
      -1,
      -2,
      3,
      -5,
      1,
      2,
      0,
      -2,
      5,
      -1,
      5,
      -5,
      0,
      2,
      -1,
      -5,
      -3,
      0,
      -28,
      2,
      -1,
      2,
      -8,
      0,
      3,
      3,
      -5,
      0,
      6,
      5,
      6,
      1,
      0,
      7,
      6,
      -7,
      4,
      -4,
      0,
      4,
      21,
      -2,
      0,
      3,
      4,
      0,
      1,
      0,
      5,
      2,
      2,
      -8,
      1,
      3,
      -4,
      5,
      1,
      0,
      2,
      -1,
      -9,
      0,
      14,
      0,
      -2,
      -3,
      0,
      1,
      0,
      -3,
      0,
      0,
      4,
      0,
      -2,
      -2,
      -2,
      -6,
      2,
      -6,
      -1,
      6,
      4,
    ]),
  ])
# ---
# name: test_search
  list([
    tuple(
      "What is the pet's name?",
      list([
        'The name of the dog is Kevin',
      ]),
    ),
    tuple(
      'What color is the house?',
      list([
        'The house is blue',
      ]),
    ),
    tuple(
      'What day is it?',
      list([
        'It is Tuesday',
      ]),
    ),
  ])
# ---
